pkg/scheduler/framework/plugins/nodeports/node_ports.go

```diff
@@ -112,11 +114,69 @@ func getPreFilterState(cycleState *framework.CycleState) (preFilterState, error)
-		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.Update}},
+		{Event: framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Delete}, QueueingHintFn: pl.isSchedulableAfterPodDeleted},
+		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterNodeChange},
 	}
 }
 
+// isSchedulableAfterPodDeleted is invoked whenever a pod deleted. It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *NodePorts) isSchedulableAfterPodDeleted(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
+	deletedPod, _, err := util.As[*v1.Pod](oldObj, nil)
+	if err != nil {
+		return framework.QueueAfterBackoff, err
+	}
+
+	// If the deleted pod is unscheduled, it doesn't make the target pod schedulable.
+	if deletedPod.Spec.NodeName == "" {
+		logger.V(4).Info("the deleted pod is unscheduled and it doesn't make the target pod schedulable", "pod", pod, "deletedPod", deletedPod)
+		return framework.QueueSkip, nil
+	}
+
+	// Get the used host ports of the deleted pod.
+	usedPorts := make(framework.HostPortInfo)
+	for _, container := range deletedPod.Spec.Containers {
+		for _, podPort := range container.Ports {
+			if podPort.HostPort > 0 {
+				usedPorts.Add(podPort.HostIP, string(podPort.Protocol), podPort.HostPort)
+			}
+		}
+	}
+
+	// If the deleted pod doesn't use any host ports, it doesn't make the target pod schedulable.
+	if len(usedPorts) == 0 {
+		return framework.QueueSkip, nil
+	}
+
+	// Construct a fake NodeInfo to check whether the target pod fits. If it fits, it means
+	// that the deleted pod doesn't make the target pod schedulable, so we can skip the event.
+	nodeInfo := framework.NodeInfo{UsedPorts: usedPorts}
+	if Fits(pod, &nodeInfo) {
+		logger.V(4).Info("the deleted pod doesn't make the target pod schedulable.", "pod", pod, "deletedPod", deletedPod)
+		return framework.QueueSkip, nil
+	}
+
+	logger.V(4).Info("an other pod, which used hostports are conflicted with the target pod, was deleted", "pod", klog.KObj(pod), "deletedPod", klog.KObj(deletedPod))
+	return framework.QueueAfterBackoff, nil
+}
+
+// isSchedulableAfterNodeChange is invoked whenever a node added or changed. It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *NodePorts) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
+	originalNode, modifiedNode, err := util.As[*v1.Node](oldObj, newObj)
+	if err != nil {
+		return framework.QueueAfterBackoff, err
+	}
+
+	if originalNode == nil {
+		logger.V(4).Info("node was created, pod may be schedulable now", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+		return framework.QueueImmediately, nil
+	}
+
+	logger.V(4).Info("node was updated", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
```

Comment:
@sanposhiho 
when we recieve a node's update event, how to decision whether we should return QueueAfterBackoff or QueueSkip? 

If we return QueueAfterBackoff, Maybe we don't need to register the `isSchedulableAfterNodeChange` func because it always return QueueAfterBackoff?

Reply:
Yes, I was wondering the same thing -- probably the Node update events won't make the Pod schedulable, right?
Then, let's:
- remove framework.Update from L118.
- remove `isSchedulableAfterNodeChange` since it'll be always returning QueueAfterBackkoff

Reply:
@sanposhiho 

@ahg-g add the update event into the nodeport plugin. https://github.com/kubernetes/kubernetes/pull/109442/files#diff-086ba145de7f62756b7be52f167ed73eeb9e70855cfed5554c7c7ffe40fbe79e

Perhaps, we still need the update event.

Reply:
Actually, it is a scary problem since most reviewers (including me) probably don't know the fact that we may miss NodeCreated events in that scenario. I need to revisit all QueueingHint again to make sure missing NodeCreated events never cause the problem 😓

Reply:
@sanposhiho added.

Reply:
And, can we remove `isSchedulableAfterNodeChange` either way? It always return QueueAfterBackoff.

Reply:
@sanposhiho  I think so. updated.